### PR TITLE
applications: serial_lte_modem: FTP features request

### PR DESCRIPTION
**Pull request by junqingzou**
Add new FTP command "uput"(unique) and "mput"(append).
Add data mode support to FTP "put", "uput" and "mput".
Add "verbose" command to turn on|off FTP control message.
Send URC for socket errors returned from ft_client lib.

Check JIRA ticket for more details.
JIRA Reference: NCSIDB-389

This PR contains 3 commits.

 Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

### CODE REVIEW

File: subsys/net/lib/ftp_client/src/ftp_client.c [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /*
  * Copyright (c) 2020 Nordic Semiconductor ASA
  *
- * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
```
Comment (rlubos) [inline subsys/net/lib/ftp_client/src/ftp_client.c:4]: Please, do not revert back to the old license.

File: subsys/net/lib/ftp_client/src/ftp_client.c [hunk 2]
```diff
@@ -99,37 +100,73 @@ static int establish_data_channel(const char *pasv_msg)
...
-	return data_sock;
+	return ret;
+}
+
+static void close_connection(int code, int error)
+{
+	if (FTP_PROPRIETARY(code)) {
+		switch (code) {
+		case FTP_CODE_901:
+			sprintf(ctrl_buf, "901 Disconencted(%d).\r\n", error);
```
Comment (rlubos) [inline subsys/net/lib/ftp_client/src/ftp_client.c:141]: typo: Disconnected

File: subsys/net/lib/ftp_client/src/ftp_commands.h [hunk 3]
```diff
@@ -120,7 +68,7 @@ extern "C" {
 /* Remove a directory */
 #define CMD_RMD		"RMD %s\r\n"
 /* Remove a directory tree */
-#define CMD_RMDA	"RMDA\r\n"
+//#define CMD_RMDA	"RMDA\r\n"
```
Comment (rlubos) [inline subsys/net/lib/ftp_client/src/ftp_commands.h:71]: Please remove commented code. Or use C-style comments if you want to keep it for any reason.

File: applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c [hunk 4]
```diff
@@ -401,51 +459,245 @@ static int do_ftp_get(void)
...
+	}
+
+	if (datatype == DATATYPE_ARBITRARY) {
+#if defined(CONFIG_SLM_DATAMODE_HWFC)
+		if (!check_uart_flowcontrol()) {
+			LOG_ERR("Data mode requires HWFC.");
+			return -EINVAL;
+		}
+#endif
+		// enter data mode
```
Comment (rlubos) [inline applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c:517]: Please use C-style comments.

File: applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c [hunk 5]
```diff
@@ -401,51 +459,245 @@ static int do_ftp_get(void)
...
```
Comment (rlubos) [inline applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c:591]: Please use C-style comments.

File: applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c [hunk 6]
```diff
@@ -401,51 +459,245 @@ static int do_ftp_get(void)
...
+		return ret;
+	}
+	if (datatype == DATATYPE_ARBITRARY) {
+#if defined(CONFIG_SLM_DATAMODE_HWFC)
+		if (!check_uart_flowcontrol()) {
+			LOG_ERR("Data mode requires HWFC.");
+			return -EINVAL;
+		}
+#endif
+		// enter data mode
```
Comment (rlubos) [inline applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c:664]: Please use C-style comments.

**Review by ru-fu — APPROVED** [review]
Doc update looks good.

File: subsys/net/lib/ftp_client/src/ftp_client.c (hunk 1 above)
Comment (junqingzou) [reply subsys/net/lib/ftp_client/src/ftp_client.c:4]: updated

File: subsys/net/lib/ftp_client/src/ftp_commands.h (hunk 3 above)
Comment (junqingzou) [reply subsys/net/lib/ftp_client/src/ftp_commands.h:71]: deleted

File: applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c (hunk 4 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c:517]: updated

File: applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c (hunk 6 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c:664]: updated

File: applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c (hunk 5 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/ftp_c/slm_at_ftp.c:591]: updated

File: subsys/net/lib/ftp_client/src/ftp_client.c (hunk 2 above)
Comment (junqingzou) [reply subsys/net/lib/ftp_client/src/ftp_client.c:141]: updated

**Comment by junqingzou** [comment]
DNM due to this PR is to be tested by customer first, and possible more feature request.

**Review by rlubos — APPROVED** [review]

**Comment by junqingzou** [comment]
> DNM due to this PR is to be tested by customer first, and possible more feature request.

Dear reviewers, please start the review process.

**Review by martiles — APPROVED** [review]
LGTM

**Pull request merged**